### ISSUE TITLE
what's the best practice to use multiple dataset in training? 

### ISSUE THREAD

**Issue by orcsun** [speechbrain/speechbrain#591]
first of all, thank you very much for the great work!  :P

I saw the examples are all using a single dataset.

May i ask what's the best practice to specify multiple datasets, such as librispeech and commonvoice in yaml file and merge their DynamicItemDataset?
Will it be an feature added in the future to assign dataweights to each?

many thanks!

**Comment by mravanelli** [comment]
One way is to simply concatenate the csv data manifest files generated during data preparation. I don't remember if the DynamicItemDataset already supports multiple csv files in input. Probably @popcornell and @Gastron can be more precise here.

**Comment by TParcollet** [comment]
Yes, you can use https://speechbrain.readthedocs.io/en/latest/API/speechbrain.dataio.sampler.html#module-speechbrain.dataio.sampler speechbrain.dataio.sampler.ConcatDatasetBatchSampler

**Comment by popcornell** [comment]
Exactly what @TParcollet said. 
Here is an example for TIMIT CTC ASR recipe where I am concatenating the training set and validation set. 



```python
def data_io_prep(hparams):
    "Creates the datasets and their data processing pipelines."
    data_folder = hparams["data_folder"]
    # 1. Declarations:
    train_data = sb.data_io.dataset.DynamicItemDataset.from_csv(
        csv_path=hparams["train_annotation"],
        replacements={"data_root": data_folder},
    )
    if hparams["sorting"] == "ascending":
        # we sort training data to speed up training and get better results.
        train_data = train_data.filtered_sorted(sort_key="duration")
        # when sorting do not shuffle in dataloader ! otherwise is pointless
        hparams["dataloader_options"]["train_shuffle"] = False
    elif hparams["sorting"] == "descending":
        train_data = train_data.filtered_sorted(
            sort_key="duration", reverse=True
        )
        # when sorting do not shuffle in dataloader ! otherwise is pointless
        hparams["dataloader_options"]["train_shuffle"] = False
    elif hparams["sorting"] == "random":
        pass
    else:
        raise NotImplementedError(
            "sorting must be random, ascending or descending"
        )
    valid_data = sb.data_io.dataset.DynamicItemDataset.from_csv(
        csv_path=hparams["valid_annotation"],
        replacements={"data_root": data_folder},
    )
    test_data = sb.data_io.dataset.DynamicItemDataset.from_csv(
        csv_path=hparams["test_annotation"],
        replacements={"data_root": data_folder},
    )
    datasets = [train_data, valid_data, test_data]
    label_encoder = sb.data_io.encoder.CTCTextEncoder()
    # 2. Define audio pipeline:
    @sb.utils.data_pipeline.takes("wav")
    @sb.utils.data_pipeline.provides("sig")
    def audio_pipeline(wav):
        sig = sb.data_io.data_io.read_audio(wav)
        return sig
    sb.data_io.dataset.add_dynamic_item(datasets, audio_pipeline)
    # 3. Define text pipeline:
    @sb.utils.data_pipeline.takes("phn")
    @sb.utils.data_pipeline.provides("phn_list", "phn_encoded")
    def text_pipeline(phn):
        phn_list = phn.strip().split()
        yield phn_list
        phn_encoded = label_encoder.encode_sequence_torch(phn_list)
        yield phn_encoded
    sb.data_io.dataset.add_dynamic_item(datasets, text_pipeline)
    # 3. Fit encoder:
    label_encoder.insert_blank(hparams["blank_index"])
    label_encoder.update_from_didataset(train_data, output_key="phn_list")
    # 4. Set output:
    sb.data_io.dataset.set_output_keys(datasets, ["id", "sig", "phn_encoded"])
    from speechbrain.data_io.sampler import ReproducibleRandomSampler, ConcatDatasetBatchSampler, RandomSampler
    from speechbrain.data_io.batch import PaddedBatch
    from torch.utils.data import ConcatDataset, DataLoader, SequentialSampler, BatchSampler
    # here i simply cat train and valid data as a toy example 
    tot_train_data = ConcatDataset([train_data, valid_data])
    samplers = [RandomSampler(x) for x in [train_data, valid_data]]
    train_loader = DataLoader(
        tot_train_data, 
        batch_sampler=ConcatDatasetBatchSampler(samplers,
                                                batch_sizes=[4, 3]),
        num_workers=6, collate_fn=PaddedBatch
    )
    valid_loader = DataLoader(
        valid_data,
        batch_size=4,
        drop_last=False,
        shuffle=False,
        collate_fn=PaddedBatch
    )
    test_loader = DataLoader(
        test_data,
        batch_size=4,
        drop_last=False,
        shuffle=False,
        collate_fn=PaddedBatch
    )
    return train_loader, valid_loader, test_loader, label_encoder
   
```
This batch sampler will always return 7 examples where the first 4 are from the training set and the last three from validation set. You can use a mask in the training loop to access the two datasets. 

 
```python
   batch_sampler=ConcatDatasetBatchSampler(samplers,
                                                batch_sizes=[4, 3])
```

**Comment by orcsun** [comment]
that's perfect! thank you all! 

i will close this ticket.
